Store sstable token ranges in scylla metadata and use to improve cleanup predicate

bhalevy opened the pull request:
If the sstable has token_ranges in scylla_metadata
use them to see if the sstable needs cleanup, or wheather
the sstable token ranges are all contained in the
sorted_owned_ranges.

This is much more accurate than just relying on the
"default" sstable token range which is using only the
first and last keys.

Performance comparison:
```
$ ./build/release/scylla perf-sstable --mode=write --testdir=testlog/release --partitions=10000 --iterations=10
Before (6f6ee5581e): 613809.81 +- 7916.01 partitions / sec (10 runs, 1 concurrent ops)
After  (ffcea7787f): 601031.85 +- 9221.37 partitions / sec (10 runs, 1 concurrent ops)
```
This translates to 2% regression

Fixes #12215

bhalevy committed:
table: set owned_ranges in configure_writer

To be used to generate the sstable's owned ranges
on memtable flush, compaction, streaming, etc.

Note that we change schema_applier::commit_tables_and_views
and schema_applier::commit_on_shard to be async.
This is safe since they are called internally by
schema_applier::commit() which is already async,
and it used replica::database::lock_tables_metadata to
serialize changes.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
sstables: mx/writer: store sstable token ranges in scylla metadata

The current owned ranges are used as a hint to build
the token ranges vector for written sstables.

It is possible that tokens outside the owned ranges
would be written as we do "write both" to pending ranges,
so for example during decommission, other nodes may serve writes
to ranges they do not yet own.  Those are still "segregated"
into token ranges of their own as we start a new range once we
hit a token that is not before the next owned range.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
scylla-sstable: correctly dump sharding_metadata

This patch fixes 2 issues at one go:

First, Currently sstables::load clears the sharding metadata
(via open_data()), and so scylla-sstable always prints
an empty array for it.

Second, printing token values would generate invalid json
as they are currently printed as binary bytes, and they
should be printed simply as numbers, as we do elsewhere,
for example, for the first and last keys.

Fixes #26982

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
compaction_manager: needs_cleanup: use sstable token_ranges

If the sstable has token_ranges in scylla_metadata
use them to see if the sstable needs cleanup, or wheather
the sstable token ranges are all contained in the
sorted_owned_ranges.

This is much more accurate than just relying on the
"default" sstable token range which is using only the
first and last keys.

Fixes #12215

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

denesb commented on db/schema_applier.cc:
```diff
@@ -957,12 +957,12 @@ void schema_applier::commit_tables_and_views() {
     for (auto& schema : tables.created) {
         auto& ks = db.find_keyspace(schema->ks_name());
-        db.add_column_family(ks, schema, ks.make_column_family_config(*schema, db), replica::database::is_new_cf::yes, _pending_token_metadata.local());
+        co_await db.add_column_family(ks, schema, ks.make_column_family_config(*schema, db), replica::database::is_new_cf::yes, _pending_token_metadata.local());
```
This makes this method non-atomic, which might be a problem, considering it has "commit" in its name.

bhalevy replied:
Maybe we should have an async lock in schema_applier to enforce atomicity.
Keeping those functions synchronous makes them susceptible to reactor stalls.

denesb replied:
Ok, I just wanted to make sure we considered this aspect and this change is safe.

tgrabiec replied:
schema_applier::commit() is async, because it's a multi-shard gather, but commit_on_shard() must not be, because we need to publish all schema and topology changes instantly on a given shard.  

The comment on schema_applier says:
```
It does atomic (per shard) application of a new schema.
```

You need to construct owned ranges in the earlier preparation phase.

\cc @nuivall

denesb commented on db/schema_applier.cc:
```diff
@@ -972,7 +972,7 @@ void schema_applier::commit_tables_and_views() {
 }
 
-void schema_applier::commit_on_shard(replica::database& db) {
+future<> schema_applier::commit_on_shard(replica::database& db) {
```
Makes method non-atomic.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -63,6 +63,8 @@ inline owned_ranges_ptr make_owned_ranges_ptr(dht::token_range_vector&& ranges)
     return make_lw_shared<const dht::token_range_vector>(std::move(ranges));
 }
 
+future<owned_ranges_ptr> maybe_make_owned_ranges_ptr(locator::effective_replication_map_ptr erm, locator::host_id host_id);
```
Why is this here, when the only user is in `replica/`?

bhalevy replied:
Can be defined there.
I can also move the definition and implementation of owned_ranges_ptr out of compaction.
It was there just because it was only used by cleanup, but it's not compaction specific really.

bhalevy commented on sstables/mx/writer.cc:
```diff
@@ -1611,11 +1626,35 @@ stop_iteration writer::consume_end_of_partition() {
+        if (_token_ranges.empty() || it->after(_current_token, cmp)) {
+            while (it->after(_current_token, cmp)) {
+                if (++it == _owned_ranges_end) {
+                    on_internal_error_noexcept(slogger, format("Owned ranges iterator exhausted for token={}", _current_token));
```
It's not about not throwing due to the function being noexcept 
The idea is to recover in production rather than throw.
In this case the internal error is logged with a backtrace and we keep running.

But given https://github.com/scylladb/scylladb/pull/26948/files#r2517694760
I'll just relax these conditions so they won't be handled as internal errors at all.

denesb commented on replica/database.hh:
```diff
@@ -489,6 +489,7 @@ private:
     schema_ptr _schema;
     config _config;
     locator::effective_replication_map_ptr _erm;
+    compaction::owned_ranges_ptr _owned_ranges;
```
Please add a test for the above scenario, to make sure it works correctly.

bhalevy committed:
compaction: move owned_ranges_ptr to replica namespace

Prepare for using it outside of compaction
for storing token ranges of each sstable.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy committed:
streaming: make_streaming_consumer: use table::configure_writer

That will set cfg.owned_ranges based on the table's
to be used when writing the streamed sstable.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
In [f5dfce3](https://github.com/scylladb/scylladb/commit/f5dfce3fc0f471c84d6976e2d89041d103d0d168):
- compaction: move owned_ranges_ptr to replica namespace
- table: set owned_ranges in configure_writer
  - moved the setting of owned_range to a new entry point: configure_writer that is used by compaction and later in the series, by streaming consumer
- dropped "compaction: set writer config owned_ranges"
- sstables: mx/writer: store sstable token ranges in scylla metadata
  - handle tokens outside of owned_ranges (written to pending ranges), including resyncing to the owned_ranges
- streaming: make_streaming_consumer: use table::configure_writer

Please do not merge as the series depends on https://github.com/scylladb/scylladb/pull/26991

tgrabiec commented on sstables/open_info.hh:
```diff
@@ -81,6 +81,8 @@ struct sstable_open_config {
     // Mimics behavior when a SSTable is streamed to a given shard, where SSTable
     // writer considers the shard that created the SSTable as its owner.
     bool current_shard_as_sstable_owner = false;
+    // Do not move the sharding metadata to the sharder, keeping it in the scylla metadata..
```
> Do not move the sharding metadata to the sharder

What moving is this referring to? The patch doesn't seem to inhibit any moving.

bhalevy commented:
In v3 ([ffcea77](https://github.com/scylladb/scylladb/commit/ffcea7787f7f4bee5f6e3e2aa2ee92d80186b5ed)):
- adjusted test_cleanup_stop
- Posted perf-sstable results - 2% degradation :-(

bhalevy commented:
The need arises from field experience where manual cleanup coming from cloud management or the user is racing with automatic cleanup that is triggered by the raft topology service and the 2nd to arrive is redoing most if not all of the work since basing needs_cleanup on the [first, last] range is highly inaccurate.

tgrabiec commented on replica/database.cc:
```diff
@@ -1073,10 +1074,22 @@ db::commitlog* database::commitlog_for(const schema_ptr& schema) {
+    if (erm->get_token_metadata().sorted_tokens().empty()) {
+        co_return nullptr;
+    }
+    co_return make_owned_ranges_ptr(co_await erm->get_ranges(host_id));
```
We could avoid this if we had erm->get_local_ranges(), which serves pre-computed data.

It would also automatically recompute on topology changes.

Also, this way we would naturally have them per-keyspace with vnodes (erm is per-keyspace) rather than per table. So less memory usage.

avikivity commented:
> The need arises from field experience where manual cleanup coming from cloud management or the user is racing with automatic cleanup that is triggered by the raft topology service and the 2nd to arrive is redoing most if not all of the work since basing needs_cleanup on the [first, last] range is highly inaccurate.

I think we're piling on complexity here.
